File: src/main/java/utils/Xor.java
package utils;

public class Xor {
    public static byte[] xorSameLength(byte[] fst, byte[] snd) {
        if (fst.length != snd.length)
            throw new IllegalArgumentException("Input byte arrays must be of the same length");

        byte[] result = new byte[fst.length];

        for (int i = 0; i < result.length; i++) {
            result[i] = (byte) (fst[i] ^ snd[i]);
        }

        return result;
    }
}

File: src/main/kotlin/utils/byteops/Xor.kt
@file:OptIn(ExperimentalUnsignedTypes::class)

package utils.byteops

infix fun UByteArray.xor(other: UByteArray): UByteArray {
    require(this.size == other.size) { "Input byte arrays must be of the same length" }
    return this.zip(other).map { it.first xor it.second }.toUByteArray()
}

infix fun UByteArray.xorByte(byte: UByte): UByteArray {
    return this xorRepeat ubyteArrayOf(byte)
}

infix fun UByteArray.xorRepeat(repeatableArray: UByteArray): UByteArray {
    val result = UByteArray(this.size)
    for (i in result.indices) {
        result[i] = (this[i] xor repeatableArray[i % repeatableArray.size])
    }

    return result
}